### ISSUE TITLE
Bump Unused CSS Limit

### ISSUE THREAD

**Issue by peixotorms** [GoogleChrome/lighthouse#9353]
**tl;dr**

a) The "Remove unused CSS" should have no reference to the critical path + defer CSS (because it doesn't allow you to pass that metric even if you do that).
b) The critical path + defer info, should be moved to the "Eliminate render-blocking resources", since there is more appropriate
c) The 2 KB limit to trigger the "Remove unused CSS" needs further discussion.

---

I've been scratching my head around the opportunity named "Remove unused CSS" under the pagespeed insights web tool, and I think there should be a discussion about this metric:

**The opportunity description**
Currently, the "Remove unused CSS" states that _Remove dead rules from stylesheets and defer the loading of CSS not used for above-the-fold content to reduce unnecessary bytes consumed by network activity._
It then points to [https://developers.google.com/web/tools/lighthouse/audits/unused-css](https://developers.google.com/web/tools/lighthouse/audits/unused-css) which basically only talks about deferring unused CSS.

These are two different things and the description (and possibly the metric itself) need changes.
If the metric is for "Remove unused CSS" then there should be no information at all, regarding deferring CSS. 
The document for the unused-css page, basically tells us to extract the critical path, inline it on the header and defer the remaining CSS code... however, doing that will not pass the "Remove unused CSS" opportunity.
The only thing that will pass the "Remove unused CSS" opportunity, is to "remove dead rules from stylesheets" (nothing to do with deferring or the critical path), so developers will spend hours trying to defer CSS and extract the critical path, while still seeing this opportunity.
All information about deferring CSS should be deleted from the "Remove unused CSS" and moved into the "Eliminate render-blocking resources", since there is where it really fits.

Leaving only an explanation to "Remove dead rules from stylesheets" under "Remove unused CSS" is self-explanatory and doesn't require a documentation page for that... however, enter the 2KB limit.

**Some background info:**
When working with wordpress sites, very often we have plenty of css files. 
We can automate the merging of CSS files using a plugin, so the merged CSS file will include both used and unused css rules on that page.

**The 2Kb limit**
The "Remove unused CSS" metric stands correct in these cases, because the resulting CSS file usually has a lot of rules which are unused on that page. However, any CSS file with more than 2KB of unused CSS code (on that page), will trigger this opportunity and affect the score significantly, in some cases.

I think that the hard limit of 2 KB to trigger this opportunity should change to a relative size, rather than an absolute value. In addition, a total network payload size for the CSS should be implemented.

How I would trigger this metric, would be something like this:

a) if the total CSS is less than 125 Kb compressed (more on this later), do nothing (pass).
b) if the total CSS is more than 125 Kb, consider a relative ratio of unused/total size.

If the CSS file is more than 125 Kb, the relative size of the unused rules should affect scores in ranges proportionally... ex: if the CSS has more than 80% of unused CSS, it should shave away more points on the metric score than if it has less than 50% and again, less than 20% unused CSS.

The reason why I defend that the metric should do nothing if the CSS is less than 125 Kb, is because sometimes more is less.

**The browser cache**

If we grab all possible CSS rules across the website (used and unused on that page), we will end up with a considerable big CSS file.
For that page alone, there is unused code, however, that file may include rules used across the website and therefore, it makes sense to be included there.

Generating a larger CSS file once and enqueueing it (async/defer), will store it on the browser cache and it can be reused across the site for multiple pageviews.
It's true that the initial network payload will be higher, but it get's diluted with more pageviews.

If developers have to remove unused CSS code from their CSS files (with only a 2KB margin), what this will cause is that every page will need it's own CSS file, which will have to be downloaded per pageview, even if some of that code is repeated.

What this causes ultimately, is that while each page may have a smaller individual network load, the amount of data downloaded if the user visits multiple pages will increase. 
If the goal was to save data for the mobile user, we need to consider that while this metric is applicable for one page designs, it doesn't really apply for large sites and for repeating pageviews across the site (unless specific scenarios). 

Moreover, even if the developer removed all unused CSS code from the site, the chances of the final CSS file having only less than 2KB of unused CSS on all pages, is nearly zero.
In order to take full advantage of browser cache, some unused css allowance (within reasonable) should be given, to pass this metric opportunity. 

The only issue here, would be the 2KB limit being too strict without considering total CSS size or the ratio of used/unused code if it goes over the limit. 

It should not matter much if 80% of 50 Kb is unused (if we know it's used somewhere else), or perhaps if 10 to 20% out of 150 Kb is unused.

In my opinion, this needs further discussion by the team because I've seen developers creating one CSS file per page, just to pass this metric, thus eliminating any browser cache performance.

Clients care more about the metric number, than the actual speed... so if the goal is speed, there is an obligation from google to discuss this rather than simply focusing on one page tests.

Thanks

**Comment by peixotorms** [comment]
@robdodson no worries, thanks

**Comment by patrickhulce** [comment]
Thanks for the well thought out explanation @peixotorms!

A few immediate thoughts:

> a) The "Remove unused CSS" should have no reference to the critical path + defer CSS (because it doesn't allow you to pass that metric even if you do that).

Agreed, the documentation is outdated. New docs have been written but we do not yet link to them. Is https://web.dev/unused-css-rules clearer to you?

> b) The critical path + defer info, should be moved to the "Eliminate render-blocking resources", since there is more appropriate

Agreed, see first response :)

> c) The 2 KB limit to trigger the "Remove unused CSS" needs further discussion.

I'm not sure I agree totally here, but there's definitely room for *some* movement. 

Perhaps I'm misinterpreting but a majority of your comments here seem to be concerned with the fact that what is flagged here affects the score, but rest assured it does not. **No opportunity result has any impact on the score**. The score is purely a function of the metrics. Opportunities are just meant as a way to guide developers to improving their score.

With that in mind, we certainly aren't trying to recommend that all the styles for each page get split into individual CSS files and nothing is shared between them. As you note, that would be inefficient. This audit is meant to flag the situation in which a developer brings in a gigantic CSS bundle only to use a handful of styles from it. Code splitting is helpful, and creating a shared CSS bundle with all the common styles and supplementing with page-specific styles is the preferred approach. 2 KB is certainly low though and the weight of 2 KB of CSS is nothing compared to say 2 KB of JS, so I'd be onboard with increasing this to a more lenient number, say 10 KB? 10 KB of compressed CSS should cover nearly 100% of reasonable user-written cases and start only flagging the egregious "bring in all of bootstrap for a single class" issues. No matter what, the estimated ms savings for this audit should typically be quite low and relatively low priority compared to the other opportunities.

**Comment by connorjclark** [comment]
> No opportunity result has any impact on the score.

We should make this more clear in the report. It comes up a lot.

**Comment by brendankenny** [comment]
> We should make this more clear in the report. It comes up a lot.

#8717!

**Comment by peixotorms** [comment]
@patrickhulce now you got me... but I have a few more though as well.

a) Yes, https://web.dev/unused-css-rules is clearer for the "Remove unused CSS" opportunity, however the last part "Inline critical CSS & defer non-critical CSS" should not be there (rather should be on the "Eliminate render-blocking resources").

b) Great that you agree, as explained in a) 

c) I been working with Pagespeed metrics for a long time, and I didn't knew that the opportunities won't affect the scores. I think majority of people sees those opportunities as a way to improve their score, therefore they will implement whatever they can hoping for the scores to go up.

Even if the opportunities don't directly affect scores, opportunities state that "These optimizations can speed up your page load" which will of course, will change score on the "Lab Data" section, from which scores are then calculated.

For example, here is an example for a page that scores 98 on mobile:
http://prntscr.com/odyhrx

And the only opportunity is CSS:
http://prntscr.com/odyigm

If I am to remove that extra 28 KB of potential savings from the CSS, I'm guessing that the score will probably go up to 100, therefore, it may not affect the score directly, but it affects the metrics in which the score is built upon.

Then perhaps if I stand correct, each opportunity should not only have an "estimated savings time" column, but also which metric out of the Lab Data is affecting.

In addition, each metric out of the Lab Data should have an explanation of the percentage, each metric is contributing to the final score. And again, each metric, should have a range as well, rather than just red and green.

Looking at the docs here:
https://github.com/GoogleChrome/lighthouse/blob/master/docs/scoring.md#how-are-the-scores-weighted

```
The metric results are not weighted equally. Currently the weights are:

3X - first contentful paint
1X - first meaningful paint
2X - first cpu idle
5X - time to interactive
4X - speed index
0X - estimated input latency (I think, has been renamed to Max Potential First Input Delay)
```

First... if 0X means zero, it should not be there.
Also from what I understand, it would be possible to have a lower time to interactive, balancing out for slow times on other metrics?

I would like to see exactly, how many points out of the final score am I getting per metric AND how many points per metric can I get, so I would know where to focus.

For example, something like this:
You got 100/100 points on first cpu idle and it contributed 10 points to your score.
You got 100/100 points on first meaningful paint and it contributed 20 points to your score.

There would be a range in milliseconds in which google decides its enough to achieve 100% score for each metric separately and how much each opportunity will contribute for that individual score.

I understand it's a bit of work ahead to elaborate so much information, but for the sake of continuous improvement, I think this could be at least discussed and added to the roadmap.

---

@brendankenny thanks for the reference, which lead me to 
`https://github.com/GoogleChrome/lighthouse/blob/master/docs/scoring.md#how-are-the-scores-weighted`

The scores on Pagespeed Insights web tool, link to https://developers.google.com/web/tools/lighthouse/v3/scoring

I think, it's a matter of updating the documentation with extra info from https://github.com/GoogleChrome/lighthouse/blob/master/docs/scoring.md#how-are-the-scores-weighted

**Comment by patrickhulce** [comment]
Ideally, yes, we could perfectly predict how many points the score would increase for each opportunity, but in practice it's not that simple. Overall, I'm concerned that these changes in aggregate would create an even more score obsessed report than we have right now. The score is a helpful tool in incentivizing action, but orienting 100% of information in the report around the score IMO goes too far.

A few more specific responses below:

> however the last part "Inline critical CSS & defer non-critical CSS" should not be there (rather should be on the "Eliminate render-blocking resources").

We need to at least mention it even if it's not the focus because some rules can appear "unused" if they're just used on interaction. Point taken about having the majority of content in "Eliminate render-blocking resources" but no mention of this category of CSS at all is probably not on the table.

> If I am to remove that extra 28 KB of potential savings from the CSS, I'm guessing that the score will probably go up to 100, therefore, it may not affect the score directly, but it affects the metrics in which the score is built upon.

Only sometimes. The savings could decrease TTI but if the last point you need is on FCP then it won't move the score.

> each opportunity should not only have an "estimated savings time" column, but also which metric out of the Lab Data is affecting.

This has been discussed before and we're considering it, but there's a lot of potential for user confusion here and information overload, so we're hesitant.

> And again, each metric, should have a range as well, rather than just red and green.

I don't understand what you mean by this.

> There would be a range in milliseconds in which google decides its enough to achieve 100% score for each metric

There already is under the hood, it just isn't exposed.

> I understand it's a bit of work ahead to elaborate so much information

My primary concern isn't with the amount of work on our side, it's on the amount of work for new users to make sense of what's happening in the report. To fully adhere to your suggestions we would need a matrix of opportunities, many would be duplicated, the amount of information in that section would quintuple, etc. The report is teetering on the edge of information overload already and such a massive increase seems likely to push it over.
